### PR TITLE
Add asynchronous planning and execution

### PR DESCRIPTION
**Pull request by amalnanavati**
# Description

The primary goal of this PR is to enable users of `pymoveit2` to plan and/or execute asynchronously. This is useful to allow users to cancel planning/execution calls, and to allow the main thread to continue executing during planning/execution (e.g., we use `pymoveit2` within a [behavior tree framework](https://py-trees.readthedocs.io/en/devel/), so the tree must be able to continue ticking as the robot is moving). 

Changes 1-2 below tackle that primary goal. As part of tackling the primary goal, some additional changes also made sense; those are Changes 3-4, which are justified below.

## Detailed Changes

1. Make action execution (both `MoveGroup` and controller execution) asynchronous. Track the state of the action with a new enum, `MoveIt2State`, and allow users to query the state and cancel execution.
2. Carve out the asynchronous planning logic into another function, `plan_async`, that users of this API can call directly. Create an affiliated `get_trajectory` function that processes the future and returns a trajectory.
3. Make all invocations of `plan`/`plan_async` use the planning service (earlier users could decide whether to use the service or the MoveGroup action with `plan_only=true`).
    1. The reason for this is that the way of interacting with asynchronous actions call is very different from asynch services, since action calls first return a goal handle, whereas service calls directly return a future for the result. This would make the return value of `plan_async`, and how users interact with that return value, quite different depending on whether they invoke the action or the service.
    2. Since this code already has a way to asynchronously call, query, and cancel a `MoveGroup` action (via Change 1 above), I think a better way to re-enable planning-only via MoveGroup would be to allow users to set the `plan_only` property of the MoveGroup goal, and then allow them to invoke `_send_goal_async_move_action` (which is invoked anyway if they use `move_to_pose` or `move_to_configuration`). This is unimplemented — let me know if you’d like it.
4. Replace the `FollowJointTrajectory` action exposed by the controller with the `ExecuteTrajectory` action exposed by MoveIt2.
    1. This is necessary to allow us to cancel the goal via [MoveIt2’s `/trajectory_execution_event` topic](https://github.com/ros-planning/moveit2/blob/0e1e3768b05c370202ec697dfbc8d3718094311a/moveit_ros/planning/trajectory_execution_manager/src/trajectory_execution_manager.cpp#L46). Although we could alternatively cancel the action through the goal handle, the action server is able to reject the cancellation request. This topic provides us a way to more directly stop execution, but it can only be done if we execute using MoveIt’s `ExecuteTrajectory` action.
    2. I don’t think any features are lost with this switch, since MoveIt’s `ExecuteTrajectory` internally calls the controller’s `FollowJointTrajectory` action.

# Testing

- [x] Launch the [simulated panda arm](https://github.com/AndrejOrsula/panda_ign_moveit2): `ros2 launch panda_moveit_config ex_fake_control.launch.py`
- [x] Test `move_to_configuration`:
    - [x] Test the synchronous execution still works: `ros2 run pymoveit2 ex_joint_goal.py --ros-args -p joint_positions:="[1.57, -1.57, 0.0, -1.57, 0.0, 1.57, 0.7854]"`
    - [x] Restore to initial joint positions: `ros2 run pymoveit2 ex_joint_goal.py --ros-args -p joint_positions:="[0.0, -0.7853981633974483, 0.0, -2.356194490192345, 0.0, 1.5707963267948966, 0.7853981633974483]"`
    - [x] Test the asynchronous execution can be successfully canceled: `ros2 run pymoveit2 ex_joint_goal.py --ros-args -p joint_positions:="[1.57, -1.57, 0.0, -1.57, 0.0, 1.57, 0.7854]" -p synchronous:=False -p cancel_after_secs:=1.0`
    - [x] Restore to initial joint positions
    - [x] Test that asynchronous execution succeeds if not canceled: `ros2 run pymoveit2 ex_joint_goal.py --ros-args -p joint_positions:="[1.57, -1.57, 0.0, -1.57, 0.0, 1.57, 0.7854]" -p synchronous:=False -p cancel_after_secs:=-1.0`
- [x] Test `move_to_pose`:
    - [x] Restore to initial joint positions
    - [x] Test the synchronous execution still works: `ros2 run pymoveit2 ex_pose_goal.py --ros-args -p position:="[0.25, 0.0, 1.0]" -p quat_xyzw:="[0.0, 0.0, 0.0, 1.0]" -p cartesian:=False`
    - [x] Restore to initial joint positions
    - [x] Test the asynchronous execution can be successfully canceled: `ros2 run pymoveit2 ex_pose_goal.py --ros-args -p position:="[0.25, 0.0, 1.0]" -p quat_xyzw:="[0.0, 0.0, 0.0, 1.0]" -p cartesian:=False -p synchronous:=False -p cancel_after_secs:=1.0`
    - [x] Restore to initial joint positions
    - [x] Test that asynchronous execution succeeds if not canceled: `ros2 run pymoveit2 ex_pose_goal.py --ros-args -p position:="[0.25, 0.0, 1.0]" -p quat_xyzw:="[0.0, 0.0, 0.0, 1.0]" -p cartesian:=False -p synchronous:=False -p cancel_after_secs:=-1.0`
- [x] Test `gripper`:
    - [x] `ros2 run pymoveit2 ex_gripper.py --ros-args -p action:="open"`
    - [x] `ros2 run pymoveit2 ex_gripper.py --ros-args -p action:="close"`
    - [x] `ros2 run pymoveit2 ex_gripper.py --ros-args -p action:="toggle"`

Further, for our application we have developed [behavior tree behaviors](https://github.com/personalrobotics/ada_feeding/tree/ros2-devel/ada_feeding/ada_feeding/behaviors/moveit2) that use this PR’s changes to asynchronously plan and execute (with `use_move_group_action=False`). We’ve been using those behaviors for months and they have worked reliably.

### CODE REVIEW

**Comment by amalnanavati** [comment]
(Note: I think squashing and merging would make the most sense given the many commits)

File: examples/ex_joint_goal.py [hunk 1]
```diff
@@ -33,6 +35,9 @@ def main():
             0.7853981633974483,
         ],
     )
+    node.declare_parameter("synchronous", True)
+    # If negative, don't cancel. Only used if synchronous is False
+    node.declare_parameter("cancel_after_secs", -1.0)
```
Comment (AndrejOrsula) [inline examples/ex_joint_goal.py:40]: I think that setting this parameter to `0.0` should also **not** cancel the execution. 
Would you consider changing this parameter to have an effect only for positive numbers and set the default to `0.0` (also in the other example)?

File: pymoveit2/moveit2.py [hunk 2]
```diff
@@ -48,10 +63,9 @@ def __init__(
         base_link_name: str,
         end_effector_name: str,
         group_name: str = "arm",
-        execute_via_moveit: bool = False,
+        use_move_group_action: bool = False,
         ignore_new_calls_while_executing: bool = False,
         callback_group: Optional[CallbackGroup] = None,
-        follow_joint_trajectory_action_name: str = "joint_trajectory_controller/follow_joint_trajectory",
```
Comment (AndrejOrsula) [inline pymoveit2/moveit2.py:54]: Are these parameters the only breaking changes introduced by this PR for the public API?

If so, do you think it would be possible to keep the original parameters while indicating that they are *deprecated* to prevent sudden breaking changes? If it is possible, it would be great!

For example, `follow_joint_trajectory_action_name` could potentially be ignored, and the meaning of `execute_via_moveit` could be combined with the new `use_move_group_action`. If they are set to non-default values, then a deprecation warning would be printed.

**Review by AndrejOrsula — COMMENTED** [review]
Thank you very much for your contribution. It looks great!

I have two comments. One of them discusses the breaking changes of this PR. If it would be possible to hide/remove these breaking changes, it would be much simpler to merge this PR into the primary branch. Regardless, it is probably still better to merge it into `devel` as a beginning to give some time for testing.

File: pymoveit2/moveit2.py (hunk 2 above)
Comment (amalnanavati) [reply pymoveit2/moveit2.py]: While these parameters are the only explicit breaking changes to the public API, there is an additional **implicit** change because execution now goes through the `/execute_trajectory` action (of type `moveit_msgs/ExecuteTrajectory`) and not the action specified `follow_joint_trajectory_action_name` (of type `control_msgs/FollowJointTrajectory`). In cases where the user has configured their controller with `moveit2`, that should be fine. But in the off-chance that they have not configured their controller with `moveit2`, then this class will no longer be able to execute their robot.

I considered adding an either-or system where users can either use `ExecuteTrajectory` or use `FollowJointTrajectory`. However, the issue is that cancellation logic is very different for each. Whereas `ExecuteTrajectory` can be canceled via topic publication, `FollowJointTrajectory` requires making a cancellation request to the action server, which can be accepted or rejected. That’s why I ended up removing execution via `FollowJointTrajectory` altogether, in favor of execution via `ExecuteTrajectory`. I think that also makes more sense, because this class is an interface for `moveit2`, so it makes sense for execution to also go through `moveit2`.

Re. your changes about deprecation, I made the changes, but I don’t think the check for non-default values will work well for `follow_joint_trajectory_action_name `. Because some existing users may not pass in a custom value because the default value is the true name for their controller’s action. For those users, and the deprecation warning won’t show, so they won’t know why commands aren’t being sent to the controller.

**Comment by amalnanavati** [comment]
Thanks for the prompt reply! I addressed all the comments, rebased onto and ran the pre-commit hook, and re-tested it. I think there is slight further discussion to be had on breaking changes (see comment).

Just FYI, there will be more PRs coming over the coming days/weeks that build off of this. My lambates and I have been making multiple feature additions in our fork over the last few months, including expanding the API to allow for path constraints, modify the allowed collision matrix, move collision boxes, scale collision meshes, etc. I'll create PRs for them one-by-one, and target `devel` as this PR does.

File: pymoveit2/moveit2.py (hunk 2 above)
Comment (AndrejOrsula) [reply pymoveit2/moveit2.py]: > there is an additional **implicit** change because execution now goes through the `/execute_trajectory` action (of type `moveit_msgs/ExecuteTrajectory`) and not the action specified `follow_joint_trajectory_action_name` (of type `control_msgs/FollowJointTrajectory`). 

> ... cancellation logic is very different for each. Whereas `ExecuteTrajectory` can be canceled via topic publication, `FollowJointTrajectory` requires making a cancellation request to the action server, which can be accepted or rejected. That’s why I ended up removing execution via `FollowJointTrajectory` altogether, in favor of execution via `ExecuteTrajectory`. I think that also makes more sense, because this class is an interface for `moveit2`, so it makes sense for execution to also go through `moveit2`.

I think you are right, and this change should be fine.

> I considered adding an either-or system where users can either use `ExecuteTrajectory` or use `FollowJointTrajectory`. 

As long as `ExecuteTrajectory` covers everything that's required for basic usage with simple arms and grippers, then sticking to this interface should be fine.

> Re. your changes about deprecation, I made the changes, but I don’t think the check for non-default values will work well for `follow_joint_trajectory_action_name `. Because some existing users may not pass in a custom value because the default value is the true name for their controller’s action. For those users, and the deprecation warning won’t show, so they won’t know why commands aren’t being sent to the controller. So fi we keep `follow_joint_trajectory_action_name`, perhaps we should print the deprecation warning every time? But that will be noisy for folks who don't use that parameter...

Thank you. I think it is fine. Alternatively, the default value for `follow_joint_trajectory_action_name` could be set to a special "DEPRECATED" string, an empty string or even None. It does not really matter what the default is if it is not used anyway. :)

**Review by AndrejOrsula — COMMENTED** [review]
Thank you for addressing those changes!

---

I just realized that this PR breaks  [MoveIt2Gripper](https://github.com/AndrejOrsula/pymoveit2/blob/6268b9993352b408e268b18856792deecb1bcffa/pymoveit2/moveit2_gripper.py) and in turn [GripperInterface](https://github.com/AndrejOrsula/pymoveit2/blob/6268b9993352b408e268b18856792deecb1bcffa/pymoveit2/gripper_interface.py) because `MoveIt2Gripper` inherits from `MoveIt2` and reuses some of its internals (including the replaced `FollowJointTrajectory` action client).

E.g. running [ex_gripper.py](https://github.com/AndrejOrsula/pymoveit2/blob/6268b9993352b408e268b18856792deecb1bcffa/examples/ex_gripper.py):
```
'GripperInterface' object has no attribute '_MoveIt2__follow_joint_trajectory_action_client'
```

Would you be interested in updating this interface accordingly? Alternatively, we can merge this into `devel` now and then address grippers in some upcoming PR.

**Comment by amalnanavati** [comment]
Fixed the gripper interface, and updated the tests accordingly. Note that because I don't actively use the gripper I probably didn't test it as thoroughly as can be -- I just ran the three commands at the top `ex_gripper.py`. Feel free to test more if there are further edge cases you often encounter when using the Gripper Interface :)

**Review by AndrejOrsula — APPROVED** [review]
It works well from my initial tests. Thank you very much for your contribution! :smile:
